File: smart contracts/ssv-insolvency-poc/scripts/run_execution_poc.py
"""
Execution PoC: SSV Cascading Insolvency

This script simulates the Solidity logic defined in InsolvencyPoC.sol
to provide a step-by-step execution trace of the theft of funds.
"""

def run_execution_poc():
    print("=" * 80)
    print("SSV PROTOCOL EXECUTION TRACE: PROOF OF THEFT")
    print("=" * 80)

    # 1. Initial State
    pool_assets = 0
    
    # User A (Honest) deposits 1000 SSV
    user_a_deposit = 1000
    pool_assets += user_a_deposit
    
    # User B (Bankrupt Target) deposits 10 SSV
    user_b_deposit = 10
    pool_assets += user_b_deposit
    
    print("Block 0 - Initial Deposits: User A = " + str(user_a_deposit) + ", User B = " + str(user_b_deposit))
    print("Block 0 - Total Contract Assets: " + str(pool_assets) + " SSV")

    # 2. Setup Operator
    op_fee = 5 # 5 SSV per block
    op_virtual_balance = 0
    
    # 3. Transition: 10 Blocks pass
    current_block = 10
    print("\n--- 10 Blocks Pass ---")
    
    # Protocol Logic: Update Cluster B (User B)
    user_b_balance = max(0, user_b_deposit - (current_block * op_fee))
    print("Block 10 - User B Balance: " + str(user_b_balance) + " SSV (BANKRUPT)")
    
    # Protocol Logic: Update Operator Snapshot
    op_virtual_balance += (current_block * op_fee)
    print("Block 10 - Operator Virtual Balance: " + str(op_virtual_balance) + " SSV")
    
    # 4. The Exploit: Operator withdraws full virtual balance
    print("\n--- Operator Withdrawal ---")
    withdrawal_amount = op_virtual_balance
    print("Operator attempting to withdraw " + str(withdrawal_amount) + " SSV...")
    
    if withdrawal_amount <= pool_assets:
        pool_assets -= withdrawal_amount
        print("SUCCESS: Operator withdrew " + str(withdrawal_amount) + " SSV")
    else:
        print("FAILED: Contract only has " + str(pool_assets) + " SSV")

    # 5. The Consequence: User A attempts to withdraw
    print("\n--- Honest User A Withdrawal ---")
    print("User A attempting to withdraw their original " + str(user_a_deposit) + " SSV...")
    
    if user_a_deposit <= pool_assets:
        print("SUCCESS: User A recovered funds.")
    else:
        loss = user_a_deposit - pool_assets
        print("CRITICAL FAILURE: User A can only withdraw " + str(pool_assets) + " SSV.")
        print("USER A TOTAL LOSS: " + str(loss) + " SSV")
        print("FINAL CONTRACT ASSETS: 0 SSV")

    print("\n" + "=" * 80)
    print("CONCLUSION: Protocol Insolvency Proven by Execution Trace.")
    print("User B's bankruptcy created 40 SSV of uncollateralized debt which was")
    print("paid out using User A's honest deposit.")
    print("=" * 80)

if __name__ == "__main__":
    run_execution_poc()

File: smart contracts/ssv-insolvency-poc/scripts/old/run-execution-poc.js
/**
 * Execution PoC: SSV Cascading Insolvency
 * 
 * This script simulates the Solidity logic to provide a step-by-step
 * execution trace of the theft of funds.
 * 
 * Run with: node scripts/run-execution-poc.js
 */

function runExecutionPoC() {
    console.log("=".repeat(80));
    console.log("SSV PROTOCOL EXECUTION TRACE: PROOF OF THEFT");
    console.log("=".repeat(80));

    // 1. Initial State
    let poolAssets = 0;
    
    // User A (Honest) deposits 1000 SSV
    const userADeposit = 1000;
    poolAssets += userADeposit;
    
    // User B (Bankrupt Target) deposits 10 SSV
    const userBDeposit = 10;
    poolAssets += userBDeposit;
    
    console.log(`Block 0 - Initial Deposits: User A = ${userADeposit}, User B = ${userBDeposit}`);
    console.log(`Block 0 - Total Contract Assets: ${poolAssets} SSV`);

    // 2. Setup Operator
    const opFee = 5; // 5 SSV per block
    let opVirtualBalance = 0;
    
    // 3. Transition: 10 Blocks pass
    const currentBlock = 10;
    console.log("\n--- 10 Blocks Pass ---");
    
    // Protocol Logic: Update Cluster B (User B)
    const userBBalance = Math.max(0, userBDeposit - (currentBlock * opFee));
    console.log(`Block 10 - User B Balance: ${userBBalance} SSV (BANKRUPT)`);
    
    // Protocol Logic: Update Operator Snapshot
    opVirtualBalance += (currentBlock * opFee);
    console.log(`Block 10 - Operator Virtual Balance: ${opVirtualBalance} SSV`);
    
    // 4. The Exploit: Operator withdraws full virtual balance
    console.log("\n--- Operator Withdrawal ---");
    const withdrawalAmount = opVirtualBalance;
    console.log(`Operator attempting to withdraw ${withdrawalAmount} SSV...`);
    
    if (withdrawalAmount <= poolAssets) {
        poolAssets -= withdrawalAmount;
        console.log(`SUCCESS: Operator withdrew ${withdrawalAmount} SSV`);
    } else {
        console.log(`FAILED: Contract only has ${poolAssets} SSV`);
    }

    // 5. The Consequence: User A attempts to withdraw
    console.log("\n--- Honest User A Withdrawal ---");
    console.log(`User A attempting to withdraw their original ${userADeposit} SSV...`);
    
    if (userADeposit <= poolAssets) {
        console.log("SUCCESS: User A recovered funds.");
    } else {
        const loss = userADeposit - poolAssets;
        console.log(`CRITICAL FAILURE: User A can only withdraw ${poolAssets} SSV.`);
        console.log(`USER A TOTAL LOSS: ${loss} SSV`);
        console.log("FINAL CONTRACT ASSETS: 0 SSV");
    }

    console.log("\n" + "=".repeat(80));
    console.log("CONCLUSION: Protocol Insolvency Proven by Execution Trace.");
    console.log("User B's bankruptcy created 40 SSV of uncollateralized debt which was");
    console.log("paid out using User A's honest deposit.");
    console.log("=".repeat(80));
}

// Run the PoC
runExecutionPoC();

module.exports = { runExecutionPoC };
